Collection editor

BenHenning opened the pull request:
These series of commits introduce the collection editor frontend and backend controllers. This is the MVP of the collection editor; it's fully functional and has an intentionally simplified UI.

I have reviewed all of Abe's commits, though you are free to quickly glance over them if you'd like. I don't expect you to, but you might need to when reviewing other changes and, if you find anything, feel free to leave a comment. Please focus specifically on the commits I added, since those didn't go through a thorough review.

Also, the editor can be reached by going to the following URL with the development server:

``localhost:8181/collection_editor/create/<collection_id>``

The following needs to be done yet:
- [x] Publishing collections (with proper, careful rights management with explorations).
- [x] Ensuring the collection editor is impossible to access and new collections are impossible to create except by or through whitelisted users.
- [x] File an issue for introducing a separate getter for collections for use by the editor. ``WritableCollectionBackendApiService`` should become more like ``CollectionEditorDataBackendApiService`` and the learner one should become ``CollectionViewerDataBackendApiService``.
- [x] Tests for CollectionNodeObjectFactory
- [x] Tests for CollectionObjectFactory
- [x] ~~Tests for CollectionPlaythroughObjectFactory~~
- [x] Tests for SkillListObjectFactory
- [x] Tests for ChangeObjectFactory
- [x] Tests for CollectionRightsBackendApiService
- [x] ~~Tests for WritableCollectionBackendApiService~~ (should be addressed in #1556)
- [x] Tests for UndoRedoService
- [x] Tests for CollectionUpdateService

Thanks!

seanlip commented:
@jacobdavis11 -- could you please put this through some functionality testing? Thanks!

seanlip commented on core/controllers/collection_editor.py:
```diff
@@ -0,0 +1,153 @@
+current_user_services = models.Registry.import_current_user_services()
+
+
+def _require_valid_version(version_from_payload, collection_version):
```
Any reason to make this module-level and not put it in WritableCollectionDataHandler, which I believe is the only place which uses it?

BenHenning replied:
Nope, that sounds good. Done.

seanlip replied:
Not done?

seanlip replied:
Oh, I see -- you put it back because there are now multiple handlers using it. Sorry for missing that; I retract the comment.

seanlip commented on core/domain/collection_services.py:
```diff
@@ -247,6 +247,77 @@ def is_collection_summary_editable(collection_summary, user_id=None):
+        exp_id: exp_summaries[ind] for (ind, exp_id) in enumerate(exp_ids)
+    }
+
+    # TODO(bhenning): Users should not be recommended explorations they
```
Can you file an issue for this?

BenHenning replied:
Done.

seanlip commented on core/domain/collection_services.py:
```diff
@@ -247,6 +247,77 @@ def is_collection_summary_editable(collection_summary, user_id=None):
+    for collection_node in collection_dict['nodes']:
+        summary = exp_summaries_dict.get(collection_node['exploration_id'])
+        if not summary:
+            raise utils.ValidationError(
```
So, what's going to happen from the learner UX point of view in this case? Will they see a 500 error?

I'd have thought that it'd be better for them to see something telling them that the exploration has been deleted, but they should still be able to play the rest of the collection? Or is the point that this scenario will never occur (from the learner perspective)?

BenHenning replied:
It can occur (see long explanation above). The collection is really in a bad state if an exploration is deleted, I don't see how it's possible to play it without changing the behavior of the collection for delete explorations (and even doing that isn't obvious to me). This overall process has been cleaned up quite a bit in the upcoming commit, but there's still room for improvement. I'm inclined to update the exploration deletion process and stop it from being deleted if it's part of a collection.

However, this sort of situation really shouldn't come up too often (ie, it should only be possible with private collections after the changes made in the upcoming commit). Because of that, I'm inclined to keep it as-is.

seanlip commented on core/domain/collection_services.py:
```diff
@@ -247,6 +247,77 @@ def is_collection_summary_editable(collection_summary, user_id=None):
+    includes progress in the collection, information about explorations
+    referenced within the collection, and a slightly nicer data structure for
+    frontend work.
+    """
```
This seems like a crucial function; maybe add some documentation about when errors are raised.

BenHenning replied:
Done. This also led me to fixing many bugs surrounding the deletion process for explorations and how they relate to collections. Now, if an exploration referenced within a collection is deleted, it will cause the collection player to issue an error. The collection editor now will properly display the collection node, but place a warning next to it to notify the creator that exploration is broken. The editor also now allows the deleted exploration to be safely removed from the collection, which then in turn fixes the collection player.

These errors then led me to wondering about public vs. private status of explorations and collections, so I added guards for those, as well. If a collection is private, it can include both private and public explorations. If it's public, it can only contain public explorations per the collections design document.

Finally, while investigating the above issue with rights, I also realized the collection editor is viewable by users without permission, even though they can't change anything. This has been changed; the collection editor will not allow users to view it who do not have permission to edit that collection, until a read-only version of the collection editor is implemented.

seanlip commented on core/templates/dev/head/collection_editor/CollectionDetailsDirective.js:
```diff
@@ -0,0 +1,93 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Directive displaying and editing a collection details. Edit
```
Directive for ...

seanlip replied:
Also, is this meant to be an editor or a viewer? If the former, maybe call it CollectionDetailsEditorDirective? If the latter, that's a bit confusing, and I don't see an isEditable option...

BenHenning replied:
Sgtm, done (both).

seanlip commented on core/templates/dev/head/collection_editor/CollectionDetailsDirective.js:
```diff
@@ -0,0 +1,93 @@
+
+/**
+ * @fileoverview Directive displaying and editing a collection details. Edit
+ * options include: chaning the title, changing the objective, changing the
```
changing

BenHenning replied:
Done.

seanlip commented on core/templates/dev/head/collection_editor/CollectionDetailsDirective.js:
```diff
@@ -0,0 +1,93 @@
+      };
+
+      $scope.updateCollectionTitle = function(newCollectionTitle) {
+        var collectionTitle = {
```
I don't understand this at all. Why are we making a dict? Is it because $scope.setCollectionTitle matches via kwarg? If so, perhaps a better name for the dict would be something like 'payload' or 'collectionTitlePayload'? It seems confusing to call it a collectionTitle when it isn't.

Ditto below.

seanlip replied:
(Oh, and above as well, for addExploration -- I think? I got the impression that newCollectionNode was the dict representing a collection node...but that may not be the case after all.)

BenHenning replied:
Done.

seanlip commented on core/templates/dev/head/collection_editor/CollectionDetailsDirective.js:
```diff
@@ -0,0 +1,93 @@
+          expId: newExpId
+        };
+        if ($scope.addCollectionNode(newCollectionNode)) {
+          var change = CollectionUpdateService.buildAddCollectionNodeUpdate(
```
buildAddCollectionNodeChangeDict might be a better name, otherwise we have duplicate terminology for the same thing ('Change' and 'Update'). Mildly prefer 'ChangeDict' to both of these, since it makes clear that it's a noun.

seanlip replied:
(Ditto below)

BenHenning replied:
I like the idea of this pattern. Done.

seanlip commented on core/templates/dev/head/collection_editor/CollectionEditor.js:
```diff
@@ -0,0 +1,213 @@
+
+// TODO(bhenning): These constants should be provided by the backend.
+oppia.constant(
+  'COLLECTION_DATA_URL', '/collection_handler/data/<collection_id>');
```
Optional nit: COLLECTION_DATA_URL_TEMPLATE or COLLECTION_DATA_UNINTERPOLATED_URL? Ditto below. (As it is, it's not really a URL.)

BenHenning replied:
Done.

seanlip commented on core/templates/dev/head/collection_editor/CollectionEditor.js:
```diff
@@ -0,0 +1,213 @@
+        $scope.collection = collection;
+        $scope.collectionSkills = collection.skills;
+      }, function(error) {
+        // TODO(mgowano): Handle not being able to load the collection.
```
Would it make sense to group these TODOs into an issue and file that so it can be tracked? TODOs in the codebase get stale (as you've no doubt seen :-) )

BenHenning replied:
Done in #1462 and #1463.

seanlip commented on core/templates/dev/head/collection_editor/CollectionEditor.js:
```diff
@@ -0,0 +1,213 @@
+          error || 'There was an error loading the collection.');
+      });
+
+    var _arrayContainsCaseInsensitive = function(array, element) {
```
Add docstring or rename (I read this and was thinking "array contains case-insensitive what?") Hard to find a good name though. Maybe isElementInArrayIgnoringCase? (I know, not great either.)

seanlip replied:
Or maybe: isSkillInArray? You're only using this for skills, so if you add a docstring explaining the case normalization, that would become clearer.

seanlip replied:
And, if you can do that, you can s/element/skill (or skillDescription, skillString, etc.)

BenHenning replied:
Done (all, or some combination thereof).

seanlip commented on core/templates/dev/head/collection_editor/CollectionEditor.js:
```diff
@@ -0,0 +1,213 @@
+      });
+
+    var _arrayContainsCaseInsensitive = function(array, element) {
+      var lowerElement = element.toLowerCase();
```
lowercaseElement

BenHenning replied:
Done (renamed to ``lowercaseSkillName``).

seanlip commented on core/templates/dev/head/collection_editor/CollectionEditor.js:
```diff
@@ -0,0 +1,213 @@
+
+    var _arrayContainsCaseInsensitive = function(array, element) {
+      var lowerElement = element.toLowerCase();
+      for (var i = 0; i < array.length; i++) {
```
optional: could replace these 6 lines with

    return array.some(function(arrayItem) {
      return arrayItem.toLowerCase() === lowercaseElement;
    }));

BenHenning replied:
Nice suggestion, done.

seanlip commented on core/templates/dev/head/collection_editor/CollectionEditor.js:
```diff
@@ -0,0 +1,213 @@
+
+    // TODO(mgowano): Handle cases where deleting a pre-requisite skill results
+    // to unreachable explorations.
+    $scope.deletePrerequisiteSkill = function(skillIdx, expId) {
```
What's Idx -- index? If so I would just say skillIndex.

BenHenning replied:
Done.

seanlip commented on core/templates/dev/head/collection_editor/CollectionEditor.js:
```diff
@@ -0,0 +1,213 @@
+
+    // TODO(mgowano): Handle cases where deleting an acquired skill results
+    // to unreachable explorations.
+    $scope.deleteAcquiredSkill = function(skillIdx, expId) {
```
Ditto.

BenHenning replied:
Done.

seanlip commented on core/templates/dev/head/collection_editor/CollectionEditor.js:
```diff
@@ -0,0 +1,213 @@
+    };
+
+    // TODO(mgowano): Handle a case where deleting an exploration will
+    // result into unreachable exploration. At the moment this is only
```
result in other explorations being unreachable.

BenHenning replied:
Comment was removed.

seanlip commented on core/templates/dev/head/collection_editor/CollectionEditor.js:
```diff
@@ -0,0 +1,213 @@
+    // result into unreachable exploration. At the moment this is only
+    // handled when a change is committed to the backend.
+    $scope.deleteCollectionNode = function(expId) {
+      if (expId) {
```
Why would expId be null?

seanlip replied:
(Worth explaining in docstring.)

BenHenning replied:
Rather than directly addressing this, I decided to cleanup all of these mutating functions. None of them actually needed to return boolean values, so the new versions are a bit cleaner and easier to understand. It should also address your concern here.

seanlip commented on core/templates/dev/head/collection_editor/CollectionEditor.js:
```diff
@@ -0,0 +1,213 @@
+      return false;
+    };
+
+    // TODO(mgowano): Set all values for the 'exploration' key in newNode
```
I have no idea what this TODO means; could it be elaborated on? (Also, what's stopping us from just doing it?)

BenHenning replied:
This has been replaced by #1463.

seanlip commented on core/templates/dev/head/collection_editor/CollectionEditor.js:
```diff
@@ -0,0 +1,213 @@
+    };
+
+    $scope.setCollectionTitle = function(newTitle) {
+      if (newTitle) {
```
If this condition isn't met, shouldn't some error occur?

Alternatively, maybe these methods need docstrings explaining that it is the caller's responsibility to error if they return false.

BenHenning replied:
The policies of these functions have changed, so I don't think your comment applies anymore (and should be addressed now that a warning is issued to the user for many of these actions).

seanlip commented on core/templates/dev/head/collection_editor/CollectionNodeDirective.js:
```diff
@@ -0,0 +1,126 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Directive displaying and editing a collection node. Edit
```
Directive for...

Also, same question as above: is this meant to be just the edit view? If so, worth renaming.

BenHenning replied:
Done (both).

seanlip commented on core/templates/dev/head/collection_editor/CollectionNodeDirective.js:
```diff
@@ -0,0 +1,126 @@
+
+/**
+ * @fileoverview Directive displaying and editing a collection node. Edit
+ * options include: editing pre-requisite skills, editing acquired skills and
```
More like adding/deleting skills, right?

BenHenning replied:
Clarified.

seanlip commented on core/templates/dev/head/collection_editor/CollectionDetailsDirective.js:
```diff
@@ -0,0 +1,93 @@
+      setCollectionTitle: '&',
+      setCollectionCategory: '&',
+      setCollectionObjective: '&',
+      addChange: '&'
```
Here and in the other editor directives: would it be better for this to come in as a service in the controller, rather than through the $scope? That seems more natural to me -- I'd imagine the change handler service would be a singleton.

BenHenning replied:
I agree, though I'm reluctant to do it as part of this change. I feel like that is taking a step towards a general change list service for both editors, though I want to refactor this editor to include that functionality once it's also extracted from the exploration editor. Unless you have very strong preferences here (and it doesn't seem as though you do), I'm going to settle for adding a TODO to ensure this gets done as part of future refactoring work.

seanlip replied:
I have fairly strong preferences, actually; I think that the current setup is setting a bad precedent. Will it take a long time to fix? You could have a pretty minimal change list service here to start off with (and it's fine if it's collection-editor-specific).

BenHenning replied:
This has been fixed with the latest refactor. We now have an ``UndoRedoService`` that will also be usable by the exploration editor, once that is refactored.

seanlip commented on core/templates/dev/head/collection_editor/CollectionNodeDirective.js:
```diff
@@ -0,0 +1,126 @@
+    templateUrl: 'inline/collection_node_directive',
+    controller: ['$scope', 'CollectionUpdateService', function($scope,
+      CollectionUpdateService) {
+      // Uses addPrerequisiteSkill callback function to add a pre-requisite
```
These docstrings are talking way too much about implementation, IMO. How about something like:

    Adds a prerequisite skill to the frontend collection object, and also updates the changelist.

Ditto below.

BenHenning replied:
Done.

seanlip commented on core/templates/dev/head/collection_editor/CollectionSkillsListDirective.js:
```diff
@@ -0,0 +1,32 @@
+/**
+ * @fileoverview Directive for creating a list of skills (acquired and
+ * pre-requisite skills) associated with a collection.
+ * TODO (mgowano): Update the skills list as the pre-requisite and acquired list
```
This seems like a somewhat fundamental TODO?

BenHenning replied:
Mostly. Did the TODO.

seanlip commented on core/templates/dev/head/collection_editor/collection_details_directive.html:
```diff
@@ -0,0 +1,17 @@
+  <p>version: <em><[getVersion()]></em></p>
+
+  <br>
+  <input ng-model="newExpId" type="text" placeholder="New exp ID" /><button ng-click="addExploration(newExpId)">Add Exploration</button>
```
Nit: don't need /> for input tags. Can follow the style for the other inputs below.

Also, s/exp/exploration in the placeholder.

BenHenning replied:
Done (both).

seanlip commented on core/templates/dev/head/collection_editor/collection_editor.html:
```diff
@@ -0,0 +1,88 @@
+{% block content %}
+  <div ng-controller="CollectionEditor">
+    <div ng-if="collection">
+      <!-- Collection details !-->
```
Here and below: HTML comments do not generally use a ! at the end.

BenHenning replied:
Done.

seanlip commented on core/templates/dev/head/collection_editor/collection_node_directive.html:
```diff
@@ -0,0 +1,33 @@
+  <br>
+
+  <p> Acquired skills: </p>
+  <div ng-if="(getAcquiredSkills()).length > 0">
```
Nit: don't need outer parentheses.

BenHenning replied:
Done.

seanlip commented on core/templates/dev/head/collection_editor/collection_node_directive.html:
```diff
@@ -0,0 +1,33 @@
+  <div ng-if="(getAcquiredSkills()).length > 0">
+    <ul>
+      <li ng-repeat="skill in getAcquiredSkills() track by $index"><em>
+        <[skill]></em>
```
Move em tag from previous line to beginning of this line

BenHenning replied:
Done.

seanlip commented on core/templates/dev/head/domain/collection/CollectionBackendApiService.js:
```diff
@@ -103,6 +105,14 @@ oppia.factory('CollectionDataService', [
+     * collection ID with a new collection object.
+     */
+    cacheCollection: function(collectionId, collection) {
+      _collectionCache[collectionId] = collection;
```
Do you need to angular.copy()?

BenHenning replied:
Done.

seanlip commented on core/templates/dev/head/domain/collection/WritableCollectionBackendApiService.js:
```diff
@@ -0,0 +1,93 @@
+ * @author henning.benmax@gmail.com (Ben Henning)
+ */
+
+// TODO(bhenning): I think that this might be better merged with the
```
Add to issue tracker? May require TL discussion.

BenHenning replied:
Added to TL discussion for next week (or possibly the week after), instead. Doesn't seem like an "issue" to me.

seanlip commented on core/templates/dev/head/domain/collection/WritableCollectionBackendApiService.js:
```diff
@@ -0,0 +1,93 @@
+
+      return {
+        /**
+         * Updates a collection in the backend with the provided colleciton ID.
```
typo: s/colleciton/collection

BenHenning replied:
Typo 2 for 'collection' among all the reviews. :( Fixed!

seanlip commented on feconf.py:
```diff
@@ -331,7 +331,9 @@ def get_empty_ratings():
 # TODO(sll): Add all other URLs here.
-COLLECTION_DATA_URL_PREFIX = '/collectionhandler/data'
+COLLECTION_DATA_URL_PREFIX = '/collection_handler/data'
+COLLECTION_WRITABLE_DATA_URL_PREFIX = '/collection_editor/data'
```
Should this be collection_editor_handler/data (per convention to use 'handler' for all json data urls)?

BenHenning replied:
Done.

seanlip commented on core/templates/dev/head/domain/collection/WritableCollectionBackendApiService.js:
```diff
@@ -0,0 +1,93 @@
+         * to commit a change to the collection. The new collection is passed to
+         * the success callback, if one is provided to the returned promise
+         * object. Errors are passed to the error callback, if one is provided
+         * and given an error occurs. Finally, if the update is successful, the
```
Don't think you need "and given an error occurs".

BenHenning replied:
Agreed; removed that bit.

seanlip commented:
Hi @BenHenning (and @amgowano) -- looks good in general, thanks! I've done a full pass through the code and written review comments. I would wait for these to be resolved, as well as for @jacobdavis11's functionality test, before merging into develop.

seanlip commented on core/domain/collection_services.py:
```diff
@@ -247,6 +247,77 @@ def is_collection_summary_editable(collection_summary, user_id=None):
+                'deleted?)' % collection_node['exploration_id'])
+
+        collection_node['exploration'] = {
+            'id': collection_node['exploration_id'],
```
We already generate summary dicts in summary_services. Instead of duplicating the logic here, can that be reused? (Then, if we add new properties in the future, etc., we don't have to remember to add them in two different places.)

seanlip replied:
Some additional reasoning for this: ultimately, both dicts end up manifesting in the frontend as the same exploration summary tile, which makes it even more desirable to avoid skew in the backend.

BenHenning replied:
This is wonderful; I didn't know about that service. Thank you! I also updated the related function in summary_services to include the information you just mentioned (ie, that it should be used when needing information to display on summary tiles in the frontend).

jacobdavis11 commented:
I couldn't see how to create a collection. There doesn't seem to be any button available in the gallery or dashboard (and presumably one is necessary to make this functionality actually available).

I'm not sure what you mean by localhost:8181/collection_editor/create/<exploration_id> as I don't know which exploration you're referring to here. The only way I get something is by loading the test collection from the admin console and then going to localhost:8181/collection_editor/create/0, so that is what I have been looking at. Please let me know if there is other functionality I should be testing!

_1. The style of the editor looks very stark, and does not match the style used by the exploration editor.

_2. It's strange to divorce the title, category and objective at the top of the editor from the controls to change these things. I suggest having a settings tab as is done with the exploration editor. Also the "add exploration" button is logically dissimilar to the other buttons with which it is grouped.

_3. The saving functionality does not match that of the exploration editor, and it's easy to miss that changes need to be saved as the fact that there are pending changes is only visible at the bottom of the page.

_4. The category editor does not use the selector that is used elsewhere, so people are likely to enter categories incorrectly.

_5. When adding pre-requisite skills, after I type the skill and click "Add prerequisite skill for Exploration" the skill name is not clearer from the text box. The same things seems to apply to a lot of other text boxes (e.g. adding exploration ids and commit messges).

_6. When I try to save an invalid exploration I get "Warning! {"error":"Expected to have at least 1 exploration with no prerequisite skills.","code":400}" but there were no inline warnings before this. 

_7. There is no way to re-order the explorations to put the initial ones first.

_8. I got "Warning! {"error":"Some explorations are unreachable from the initial explorations: set([u'13', u'4', u'14', u'15'])","code":400}" and I'm not sure what it means. I think it occurs because I have explorations with pre-requisites that are not achievable but that is not clear. I suggest conducting an initial check to make sure that every prerequisite has at least one exploration that yields it, as that is likely to be a common problem. Then you can get into more careful checks of reachability. Also you should say which explorations are unreachable in the error message. 

_9. In the "prerequisite skills" I suggest using instead a dropdown of the available "aquired skills" as otherwise typos are going to cause major problems.

_10. When I added an exploration it was displayed as "[pJ0UzO8oD1j6]" without its exploration name. This problem doesn't seem to occur with the pre-existing explorations. 

_11. The lists of Skills near the top doesn't seem to get updated in response to adding either pre-requisite or aquired skills. 

_12. After saving there is no confirmation that the save has been successful.

BenHenning commented:
Thanks @seanlip and @jacobdavis11. Jacob, I did make a small typo in the original comment; the URL was meant to have a collection ID at the end, not exploration ID. Also, I am addressing your comments below:

(0). (No button for creating collections). Somehow this slipped through the cracks; I've add a button to show up next to the 'Create Exploration' button, but only for whitelisted users (through the admin panel).
(1), (2), (3), (4), (12). This is a minimal viable version of the collection editor; no effort was devoted to design or aesthetics. It was meant to contain all of the minimum functionality and no more. There will be follow-up changes to further improve the look, feel, and flow of the collection editor.
(5). Thanks for pointing this out; it's been fixed.
(6). There is no validation yet (#1462).
(7). Good suggestion; will log this for a later PR.
(8). Being tracked in #1462.
(9). This is a fantastic suggestion, thanks! I think it will wait until a later PR, though.
(10). This happens until you save the collection, since the summary information of an exploration is not fetched when you add a new one. This will be fixed as part of #1462.
(11). This has been fixed.

Thanks so much for your detailed analysis and feedback! I updated #1462 to include all of your comments as action items. I also filed #1464 to track the design of the collection editor.

Follow-up commits to address the issues I commented on.

BenHenning commented:
Also please be aware that in order to create new collections now, you need to add your username to a whitelist in the admin panel  ("names of users who can use the collection editor"). The 'Create Collection' button will show up next to 'Create Exploration' button, but only on the 'My Explorations' page. Also, the edit pencil will show up when viewing collections if you have access to editing them (including being whitelisted to access the collection editor).

seanlip commented:
Thanks, @BenHenning! Please let me know if you'd like another code review, or let @jacobdavis11 know if you'd like another test of the functionality.

BenHenning commented:
Per our offline discussion, I would like a follow-up review when you have an opportunity. Thanks @seanlip.

Also, @jacobdavis11 do you think the latest version of this will be usable for you to create your logic collections? There is still the problem of making collections show up in the gallery before those collections will be playable, but that's just one more PR away after this. If the editor is not quite there for you to use it, what other suggestions do you have for us to get there?

Thanks!

seanlip commented on core/domain/collection_services.py:
```diff
@@ -247,6 +248,84 @@ def is_collection_summary_editable(collection_summary, user_id=None):
+        for exp_summary_dict in exp_summary_dicts
+    }
+
+    # TODO(#1461): Users should not be recommended explorations they
```
Suggested format:

    # TODO(bhenning): Users should ... collection (see #1461).

BenHenning replied:
Done.

seanlip commented on core/templates/dev/head/collection_editor/CollectionDetailsEditorDirective.js:
```diff
@@ -0,0 +1,112 @@
+      setCollectionTitle: '&',
+      setCollectionCategory: '&',
+      setCollectionObjective: '&',
+      // TODO(bhenning): The functionality of adding changes should be provided
```
Prefer resolving this TODO prior to merge.

BenHenning replied:
Done. I decided to implement the undo/redo service discussed in the frontend refactor document, after modeling out the final collection editor. It turns out we might not need to have a change list service anymore (though we might benefit from a ``ChangeListReportingService`` for the pop-up dialog).

seanlip commented on core/templates/dev/head/collection_editor/CollectionDetailsEditorDirective.js:
```diff
@@ -0,0 +1,112 @@
+          warningsData.addWarning('Cannot add an empty exploration ID.');
+          return;
+        }
+        if ($scope.hasCollectionNode(addCollectionNodeArgs)) {
```
bit iffy to do this; what if signature diverges? Suggest you explicitly declare hasCollectionNodeArgs above.

BenHenning replied:
I think this is no longer an applicable comment after the latest changes. The pattern of passing in args in this way has been removed everywhere, I believe.

seanlip commented on core/templates/dev/head/collection_editor/CollectionEditor.js:
```diff
@@ -0,0 +1,183 @@
+    var pathnameArray = window.location.pathname.split('/');
+    for (var i = 0; i < pathnameArray.length; i++) {
+      if (pathnameArray[i] === 'create') {
+        $scope.collectionId = pathnameArray[i + 1];
```
This is already set above (line 34). Why set it here again?

BenHenning replied:
I'm actually really not sure; that shouldn't have been there. Removed.

seanlip commented on core/templates/dev/head/collection_editor/CollectionEditor.js:
```diff
@@ -0,0 +1,183 @@
+    // To be used after mutating the prerequisite and/or acquired skill lists.
+    var _resetSkillList = function() {
+      // TODO(bhenning): This can be made far more optimal once individual
+      // prerequisite and acquired skill lists are also stored in the skill list
```
I'm not sure this is a good idea? The individual lists seem to be to be more correlated with the nodes than with the global skill list.

seanlip replied:
Actually -- ignore that, sorry. I misunderstood the comment.

seanlip commented on core/templates/dev/head/domain/collection/SkillListObjectFactory.js:
```diff
@@ -0,0 +1,146 @@
+
+    // Appends another skill list to this list, where duplicate skills are
+    // ignored.
+    SkillList.prototype.concatSkillList = function(otherSkillList) {
```
call this addSkills instead; concat doesn't convey 'removing duplicates'

seanlip replied:
In fact, how is this different from the existing addSkills?

BenHenning replied:
I lean toward keeping it as ``concat``. It's concatenating a skill list (object) with the current object. ``addSkills`` and ``setSkills`` take arrays of skills, not a skill list domain object. I clarified the documentation further. Do you have any suggestions for a better name? I lean against ``addSkills`` because it makes more sense to be in the context of skill list.

Perhaps ``addSkillList``?

seanlip replied:
addSkillsFromSkillList?

BenHenning replied:
I like that; done.

seanlip commented on core/templates/dev/head/collection_editor/CollectionEditor.js:
```diff
@@ -0,0 +1,183 @@
+
+    $scope.addCollectionNode = function(explorationId) {
+      var newNode = {
+        exploration_id: explorationId,
```
why are we storing explorationId twice (here and below)?

BenHenning replied:
This was intentional to match a pattern started several PRs ago when the backed constructs the learner dict used by both the collection player and editor. This pattern seems unused (at least after the latest refactor), so I've removed it entirely.

seanlip commented:
Some minor comments, otherwise LGTM from code perspective. Thanks @BenHenning!

Once these are fixed and @jacobdavis11 has confirmed that they work fine for him, I think this can be merged.

jacobdavis11 commented:
I had a go at creating a new collection (following your instructions) and it all went smoothly except that I couldn't publish the collection, but it looks like you plan to address that anyway.

However my suggestion would be not to have anyone (e.g. me) create collections until the collection editor has been prettified and is available for all users. Otherwise I'm worried that people will see the new functionality and want to create collections of their own, and become frustrated trying to find how to do so. Thoughts?

BenHenning commented:
Thanks @seanlip. I felt like continuing to develop temporary patterns was wasteful, so this next round contains what I consider to be the final collection editor (code-wise), with no temporary patterns. I still need to add tests for the domain objects and frontend services (which will be done before merging this in or, in the very least, before publicly turning on the collection editor; I feel these tests are critical). Sorry for the refactor and large response commit, but this satisfies many goals with one commit. Please do one more round of review when convenient and if possible. There should be no more major changes after this.

Also, I introduced a new pattern in UndoRedoService.js. Please take a look at that, especially. Per the frontend refactor document, there was some goal to pull the actual JS objects out of the Angular factory/controller clauses, but that pattern hasn't had an example written for it, until now. I slightly modified the pattern from the Angular Style Guide in which is originated to be less error-prone than theirs.

@jacobdavis11: I added publishing as a follow-up commit. PTAL when possible. Also, I think the whitelist for users who can access the collection editor should satisfy your later concern. Please let me know if that's not the case.

Thanks both!

jacobdavis11 commented:
The publication works as intended, a couple of comments:
 - After publishing the "publish collection" button is still available and throws an error if clicked.
 - For non-logged in users I would change "You must login to save your progress." to something stronger to make clear that the order in which the collection should be played is not being properly displayed as we can't track where the user is.

My concern about creating collections before the option to create them is offered to all users is that lots of ordinary users will want to create collections and not be able to. I don't think the whitelist helps here since these users won't be on it and won't have any way to know about it or get themselves added to it.

seanlip commented on core/controllers/collection_viewer.py:
```diff
@@ -55,10 +54,11 @@ def get(self, collection_id):
             'can_edit': (
                 bool(self.username) and
+                self.username in (
+                    config_domain.WHITELISTED_COLLECTION_EDITOR_USERNAMES.value) and
```
FYI, Travis reports a backend linting error here

BenHenning replied:
Thanks; fixed.

seanlip commented on core/controllers/collection_editor.py:
```diff
@@ -0,0 +1,206 @@
+                collection_services.index_collections_given_ids([
+                    collection_id])
+            else:
+                raise self.InvalidInputException(
```
Mm. Not even if I'm an admin?

BenHenning replied:
Hmm, good point. Updated to allow admins to unpublish (though it's not clear how they would even do that).

seanlip commented on core/controllers/collection_editor.py:
```diff
@@ -0,0 +1,206 @@
+            if is_public:
+                try:
+                    collection.validate(strict=True)
+                    collection_services.validate_collection_public_explorations(
```
validate_exps_in_collection_are_public

BenHenning replied:
Good wording; changed.

seanlip commented on core/controllers/collection_editor_test.py:
```diff
@@ -0,0 +1,100 @@
+        """Test access to editor pages for the sample collection."""
+        whitelisted_usernames = (
+            config_domain.WHITELISTED_COLLECTION_EDITOR_USERNAMES)
+        with self.swap(
```
I think that doing it this way is a bit iffy -- I'd rather keep the config property as-is in the code, and simulate the actions needed to modify it.

Can you look at core/tests/test_utils.py -- see set_admins() and set_moderators() -- and pull part of that out into a set_config_property() method, then make use of that new method here?

BenHenning replied:
That's quite nicer and cleaner, done. Thanks for the suggestion!

seanlip commented on core/controllers/collection_editor_test.py:
```diff
@@ -0,0 +1,100 @@
+            self.assertIn('Introduction to Collections in Oppia', response.body)
+
+            # Check that non-editors cannot access the editor page. This is due
+            # to them being whitelisted.
```
not being whitelisted?

BenHenning replied:
Fixed.

seanlip commented on core/controllers/collection_editor_test.py:
```diff
@@ -0,0 +1,100 @@
+                           self.COLLECTION_ID))
+            self.assertEqual(response.status_int, 302)
+
+            # Check that it is now possible to access and edit the editor page.
```
Check that whitelisted users can access and edit the editor page.

BenHenning replied:
Done.

seanlip commented:
@BenHenning, a question: are all the issues that @jacobdavis11 brought up, and that you decided to defer for later, being tracked somewhere? (I'm referring to this comment -- https://github.com/oppia/oppia/pull/1433#issuecomment-179557912 -- don't want stuff to fall through the cracks.)

Thanks!

seanlip commented:
Thanks, @BenHenning! This looks great; I've made some comments, but I really like the general structure that you've brought in. I also agree with you that tests are critical here, especially since some of these frameworks are going to be reused.

@jacobdavis11:
  * _For non-logged in users I would change "You must login to save your progress." to something stronger to make clear that the order in which the collection should be played is not being properly displayed as we can't track where the user is._ -- Do you have a suggestion here? One could certainly be more explicit, but I feel that this isn't the place to go into too much detail. I actually think "save your progress" captures the idea quite nicely.

jacobdavis11 commented:
Something like "Log in to view your progress and see which explorations you are recommended to do next."

BenHenning commented:
Thanks @seanlip! Regarding your comment on the issues Jacob brought up earlier, I do believe #1462 logs every one that is being deferred, with additional detail. I will be adding tests and will have a follow-up message once they are added.

Thanks also @jacobdavis11! I fixed the publish button issue (it shouldn't show up when a collection is already public or after it's made public). Regarding the wording on the learner view, what about:

``Some features may be missing and your progress will not be saved unless you login.``?

I'm not sure whether it's important for logged out users to understand that collections aren't 100% operating for them. Collections sort of devolve to a list of explorations to try, though the full experience isn't available without logging in. Maybe:

``Login to save your progress and experience all features of this collection``?

jacobdavis11 commented:
I think it makes sense to let people know what features they are missing by not being logged in, so that they can decide whether they are enough to be worth logging in for. It also provides reassurance to those who refuse to log in that they will still get a reasonable experience; saying "some features may be missing" is a bit vague and alarming.

BenHenning commented:
@seanlip do you have any additional thoughts regarding the wording? Also, I added tests for the following classes:

- ChangeObjectFactory
- UndoRedoService
- CollectionUpdateService

I still have a bunch more tests to add; I've added which ones need to be added as TODO checkboxes in the main post.

seanlip commented:
I agree that "some features may be missing" is vague and alarming; that's the impression I got too. I like "Log in to save your progress and view recommended explorations." or just "Log in to save your progress."

BenHenning commented:
FYI: I finished adding the tests I intend to add. PTAL @seanlip when you get a chance. Once you're satisfied, this is going to be merged in.

I also filed #1556 to track the backend API service work. I broke it down into sufficient detail for a hopeful starter project; anyone can work on that issue, it definitely doesn't need to be me or @amgowano.

Finally, regarding the rewording of the logged-out user warning in the collection viewer. I'm going to keep it as-is, since both Sean and I are okay with the current wording. I will keep your suggestion in mind, @jacobdavis11, but won't change the wording until collections are available for playing and we can get more feedback from users and developers on their feelings on it.

Thanks again, everyone!

seanlip commented:
Hi @BenHenning, thanks! Basically LGTM, but I noticed a few things that you might want to fix before merging.

BenHenning commented:
Thanks @seanlip, I addressed your comments and fixed a couple of other bugs I found. Moving the status to LGTM since you said it's LGTM status. I will merge it once Travis returns green.
